### PR TITLE
fix aliasing of logs, read-side only

### PR DESCRIPTION
**Pull request by laurgarn**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/service/ReportService.java [hunk 1]
```diff
@@ -50,6 +58,126 @@ public ReportService(
...
+            // getReport asks to silently return empty reportmodel if not found.
+            ReporterModel reporter = this.getReport(reportUuid, "useless roundtrip string");
+            if (reporter == null) {
+                continue;
+            }
+
+            seenBuildNodeToRootReporter.put(buildNodeUuid, reporter);
+
+            for (ReporterModel subReport : reporter.getSubReporters()) {
+                UUID groupUuid = getGroupUuidFromReporter(subReport);
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/service/ReportService.java:155]: How can you extract a groupUuid from subReport task key ?
In modification-server, taskey is never an UUID ...

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]
see remarks

File: src/main/java/org/gridsuite/study/server/service/ReportService.java (hunk 1 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/service/ReportService.java:155]: Thanks for noticing that ! I push a wrong branch. One can now check https://github.com/gridsuite/network-modification-server/pull/129

File: src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java [hunk 2]
```diff
@@ -620,30 +619,40 @@ public void notifyModificationNodeChanged(UUID studyUuid, UUID nodeUuid) {
...
+        UUID defNodeId = nodeEntity.getIdNode();
+        UUID reportUuid = getReportUuid(defNodeId);
+        AbstractNode node = repositories.get(nodeEntity.getType()).getNode(defNodeId);
+        if (node.getType() == NodeType.NETWORK_MODIFICATION) {
+            NetworkModificationNode modNode = (NetworkModificationNode) node;
+            uppingRes.add(ReportingInfos.builder()
+                .definingNodeUuid(node.getId())
+                .definingNodeName(node.getName())
+                .modificationGroupUuid(modNode.getModificationGroupUuid())
+                .reportUuid(reportUuid)
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java]: Done

File: src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java [hunk 3]
```diff
@@ -620,30 +619,40 @@ public void notifyModificationNodeChanged(UUID studyUuid, UUID nodeUuid) {
...
+                .build());
+            if (!nodeOnlyReport) {
+                getUppingReportInfosFromNode(nodeEntity.getParentNode(), false, uppingRes);
+            }
+        } else {
+            uppingRes.add(ReportingInfos.builder()
+                .definingNodeUuid(node.getId())
+                .definingNodeName(node.getName())
+                .modificationGroupUuid(null)
+                .reportUuid(reportUuid)
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java]: Done

File: src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java [hunk 4]
```diff
@@ -620,30 +619,40 @@ public void notifyModificationNodeChanged(UUID studyUuid, UUID nodeUuid) {
...
 
-    private void getParentReportUuidsAndNamesFromNode(NodeEntity nodeEntity, boolean nodeOnlyReport, List<Pair<UUID, String>> res) {
-        AbstractNode node = repositories.get(nodeEntity.getType()).getNode(nodeEntity.getIdNode());
-        res.add(0, Pair.of(getReportUuid(nodeEntity.getIdNode()), node.getName()));
-        if (node.getType() == NodeType.NETWORK_MODIFICATION && !nodeOnlyReport) {
-            getParentReportUuidsAndNamesFromNode(nodeEntity.getParentNode(), false, res);
+    private void getUppingReportInfosFromNode(NodeEntity nodeEntity, boolean nodeOnlyReport,
+        List<ReportingInfos> uppingRes) {
+        UUID defNodeId = nodeEntity.getIdNode();
+        UUID reportUuid = getReportUuid(defNodeId);
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java]: Done

File: src/test/java/org/gridsuite/study/server/service/ReportServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/test/java/org/gridsuite/study/server/service/ReportServiceTest.java]: Done

File: src/main/java/org/gridsuite/study/server/service/ReportService.java [hunk 6]
```diff
@@ -50,6 +58,126 @@ public ReportService(
...
     }
 
+    /**
+     * From a sequence of ReportInfos from the deepest node to root, ask report server for reports
+     * and pack them by node/modification-group.
+     */
+    public List<ReporterModel> getReporterModels(List<ReportingInfos> uppingReportingInfos) {
+        LinkedList<UUID> uppingNeededDefNodeUuids = new LinkedList<>();
+        Map<UUID, UUID> groupToDefiningNode = new HashMap<>();
+        Map<UUID, ReportingInfos> definingNodesInfosByUuid = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/service/ReportService.java:68]: LinkedHashMap doe not provide removeLast I then use in second part of the algorithm

File: src/main/java/org/gridsuite/study/server/dto/ReportingInfos.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/dto/ReportingInfos.java]: Done

**Review by FranckLecuyer — APPROVED** [review]

**Comment by laurgarn** [comment]
Slimane version tested OK by me.

File: src/main/java/org/gridsuite/study/server/service/StudyService.java [hunk 8]
```diff
@@ -1398,21 +1400,51 @@ private void checkStudyContainsNode(UUID studyUuid, UUID nodeUuid) {
...
+        AbstractNode nodeInfos = networkModificationTreeService.getNode(nodeUuid);
+        ReporterModel reporter = reportService.getReport(nodeInfos.getReportUuid(), nodeInfos.getId().toString());
+        Map<String, Pair<List<ReporterModel>, List<ReporterModel>>> subReportersByNode = new LinkedHashMap<>();
+        reporter.getSubReporters().forEach(subReporter -> subReportersByNode.put(getNodeIdFromReportKey(subReporter), Pair.of(new ArrayList<>(), new ArrayList<>())));
+        reporter.getSubReporters().forEach(subReporter -> {
+            if (subReporter.getTaskKey().endsWith(NETWORK_MODIFICATION_TYPE_REPORT)) {
+                subReportersByNode.get(getNodeIdFromReportKey(subReporter)).getLeft().addAll(subReporter.getSubReporters());
+            } else {
+                subReportersByNode.get(getNodeIdFromReportKey(subReporter)).getRight().addAll(subReporter.getSubReporters());
+            }
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/service/StudyService.java:1433]: So, all network modifications report will be displayed first, and then all computations report after ?

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]
See remarks

File: src/main/java/org/gridsuite/study/server/service/StudyService.java (hunk 8 above)
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/service/StudyService.java:1433]: Fixed

File: src/main/java/org/gridsuite/study/server/service/StudyService.java [hunk 9]
```diff
@@ -1398,21 +1400,51 @@ private void checkStudyContainsNode(UUID studyUuid, UUID nodeUuid) {
...
+            return parentUuid.isEmpty() ? subReporters : Stream.concat(getSubReportersByNode(parentUuid.get(), false).stream(), subReporters.stream()).collect(Collectors.toList());
+        }
+    }
+
+    private List<ReporterModel> getSubReportersByNode(UUID nodeUuid) {
+        AbstractNode nodeInfos = networkModificationTreeService.getNode(nodeUuid);
+        ReporterModel reporter = reportService.getReport(nodeInfos.getReportUuid(), nodeInfos.getId().toString());
+        Map<String, Pair<List<ReporterModel>, List<ReporterModel>>> subReportersByNode = new LinkedHashMap<>();
+        reporter.getSubReporters().forEach(subReporter -> subReportersByNode.put(getNodeIdFromReportKey(subReporter), Pair.of(new ArrayList<>(), new ArrayList<>())));
+        reporter.getSubReporters().forEach(subReporter -> {
```
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/service/StudyService.java:1428]: Fixed

**Review by FranckLecuyer — APPROVED** [review]

**Pull request merged**